### PR TITLE
[exporter/datadog] Start improved hostname resolution refactor

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Starts refactor towards improving hostname resolution.

For now, we control whether to use the old/new hostname resolution with a constant at build time, since it's not ready for people to use yet. A future PR will introduce a feature flag to preview this new hostname resolution procedure before it becomes the default.

The preview hostname resolution has the following changes on this PR:

- The EC2 hostname provider always returns the instance id for compatibility with Datadog's EC2 cloud integration.

**Link to tracking Issue:** #10424

**Testing:** Refactored unit tests; need to:

- [x] Test EC2 hostname is consistent between versions

### CODE REVIEW

**Review by KSerrania — APPROVED** [review]

**Comment by mx-psi** [comment]
Ready to merge since it has a CODEOWNER approval

**Review by codeboten — APPROVED** [review]

**Pull request merged**
